### PR TITLE
Upgrade Pydantic, FastAPI, SQLModel

### PR DESCRIPTION
**Pull request by jpmckinney**
* [x] Something weird I had to change is the default value of "source_last_updated_at" in `get_award()` to `None` (instead of `""`). You can change it back to see the error that occurs. I don't know if real data will cause issues (I don't know if the tests simulate real data) – but it was weird that it needed to be changed to `None`.

----

* [x] There are still a couple instances of `.dict(` that I assume need to be `.model_dump(`, but they aren't causing warnings. Maybe they aren't covered by the tests?

----

This warning occurs:

```
tests/test_lenders.py::test_create_credit_product
tests/test_lenders.py::test_create_lender_with_credit_product
  /Users/james/.pyenv/versions/3.11.4/envs/credere-backend/lib/python3.11/site-packages/pydantic/main.py:308: UserWarning: Pydantic serializer warnings:
    Expected `decimal` but got `float` - serialized value may not be as expected
    Expected `decimal` but got `float` - serialized value may not be as expected
    Expected `str` but got `float` - serialized value may not be as expected
    Expected `decimal` but got `int` - serialized value may not be as expected
```

~~I can fix one by changing `interest_rate` to a string in `test_lenders.py`, but why is `CreditProductBase.interest_rate` a `str` in the model?~~ It's meant to be e.g. a sentence (fixture has been updated to reflect that).

~~I don't know how to fix the `decimal` warnings, because `decimal` is not JSON serializable, which I think it needs to be in this context.~~ Solution is in the linked docs immediately below.

* [ ] Check if the [recommended approach](https://docs.pydantic.dev/latest/api/standard_library_types/#decimaldecimal)  used to fix the above errors ought to be applied to all decimals
  * Added to #239 

----

These tests fail:

```
FAILED tests/test_applications.py::test_approve_application_cicle - KeyError: 'title'
FAILED tests/test_applications.py::test_application_declined - KeyError: 'status'
FAILED tests/test_applications.py::test_application_rollback_declined - KeyError: 'status'
FAILED tests/test_lenders.py::test_create_credit_product - KeyError: 'lower_limit'
FAILED tests/test_lenders.py::test_update_lender - KeyError: 'name'
FAILED tests/test_users.py::test_update_user - KeyError: 'name'
```

Basically, the API responds with `{}` instead of a non-empty dict. ~~I have no idea why.~~

* [x] Check whether other endpoints need similar changes as in https://github.com/open-contracting/credere-backend/pull/238/commits/de5d7423ae39d2b31de42fc3539d9e0ce908c5c6 Right now, tests of POST and PUT requests that only check the response code are passing. If the frontend uses the response text, then we need to test that the response is non-empty (and it will likely be empty without the changes like in the mentioned commit).

----

Note that a508184 and 1eec426 in this PR have already been merged to the main branch as different commits.

### CODE REVIEW

**Comment by jpmckinney** [comment]
As described at https://sqlmodel.tiangolo.com/tutorial/automatic-id-none-refresh/, SQLAlchemy expires objects after commit, and automatically refreshes them if attributes are accessed (e.g. `user.name`) but not if the object as a whole is used (e.g. `user)`:

> We didn't access the object's attributes, like hero.name. We only accessed the entire object and printed it, so SQLAlchemy has no way of knowing that we want to access this object's data.

The point of the refresh is mainly to get, e.g. the `id` and `created_at` of a new object, or the `updated_at` of an updated object. Otherwise, the object in memory should match the row in the table (since this all occurs in a transaction).

To refresh manually, `refresh()` needs to be called after `commit()`. This means replacing `transaction_session` with a simpler context manager and manually calling `commit()` everywhere.

The other option is to disable SQLAlchemy's expiry by setting `expire_on_commit=False`. However, I feel like this is more likely to introduce accidental bugs. I think it's safer to ensure a match between the object in memory and the row in the table. https://docs.sqlalchemy.org/en/20/orm/session_basics.html

**Comment by jpmckinney** [comment]
Okay, I got tests passing, except I uncovered #243 which errors in this PR, and the warning in the issue description is related to the other failure (where a float submitted as a decimal is returned as a str).

However, I'm not sure that all the necessary changes have been made. Many tests just check the HTTP status, but not the response format. I don't know if the frontend actually expects a response to POST and PUT requests, but if it does, and if the tests don't check the response format, it might start receiving empty responses for the endpoints that weren't touched in my latest commit.

File: app/settings.py [hunk 1]
```diff
@@ -43,51 +23,44 @@ def sentry_filter_transactions(event: dict[str, Any], hint: dict[str, Any]) -> d
...
+    days_to_erase_borrower_data: int = 7
+    days_to_change_to_lapsed: int = 14
+    ocp_email_group: str = "credere@noreply.open-contracting.org"
+    max_file_size_mb: int = 5
+    reminder_days_before_expiration: int = 3
+    progress_to_remind_started_applications: float = 0.7
+    environment: str = "development"
+    transifex_token: str = ""
+    transifex_secret: str = ""
+    model_config = SettingsConfigDict(env_file=".env", extra="allow")
```
Comment (jpmckinney) [inline app/settings.py:63]: Just curious - what are these extras that we want to allow? Should we make them explicit settings?

File: app/settings.py [hunk 2]
```diff
@@ -108,7 +81,7 @@ class Config:
             },
         },
         "loggers": {
-            "": {
+            "logger": {
```
Comment (jpmckinney) [inline app/settings.py:84]: Hmm, `""` configures the root logger, "logger" configures a logger called "logger", which I'm not sure is something we actually use?

Comment (yolile) [reply app/settings.py:84]: Ah sorry, I forgot to delete that while I was doing some testing

File: app/settings.py (hunk 1 above)
Comment (yolile) [reply app/settings.py]: Good catch, the issue was with `days_to_erase_borrowers_data` which was named `days_to_erase_borrower_data` but days_to_erase_borrowers_data in the env file. Now we don't need `extra=allow` anymore

**Comment by yolile** [comment]
> I can fix one by changing interest_rate to a string in test_lenders.py, but why is CreditProductBase.interest_rate a str in the model?

`interest_rate` is a string indeed, maybe "`interest_rate_description`" could have been a better name, as we don't put the actual interest rate there but rather a description like "The interest rate of the credit lines is variable and is subject to the Current Bank Interest for the type of consumer credit defined by the Financial Superintendency for each period.". We only learned that the FIs weren't comfortable with putting a fixed interest rate in Credere when defining the credit products before the launch.

**Comment by jpmckinney** [comment]
Cool, I fixed that one, for the decimals: https://docs.pydantic.dev/latest/api/standard_library_types/#decimaldecimal

> Serialization: Pydantic serializes Decimal types as strings. You can use a custom serializer to override this behavior if desired.

And they give an example that replaces Decimal with:

```python
    Annotated[
        Decimal,
        PlainSerializer(
            lambda x: float(x), return_type=float, when_used='json'
        ),
    ]
```

Right now, the lender tests fail because they expect a float, but the response is a `str`. Assuming that's correct, I'll change the code as above.

**Comment by yolile** [comment]
> Assuming that's correct, I'll change the code as above.

Yes, that is correct

**Comment by jpmckinney** [comment]
Okay, the code now returns a float. There is still a Pydantic warning about parsing a float/int into a decimal.

I don't know of a Pydantic-specific way to avoid this warning - maybe I didn't look hard enough.

I also don't know if Pydantic has a special JSON parser to directly load JSON numbers as decimals.

But, in the tests at least, it is safe to convert float/int to Decimal, so I just ignore the warnings.

**Comment by jpmckinney** [comment]
Update: In case Pydantic had some magic sauce per above, I tested the patch below but it has warnings anyway. In a [related issue](https://github.com/pydantic/pydantic/issues/6467#issuecomment-1623680485), Pydantic suggests ignoring the warning (though I think the linked issue is narrower than our issue).

<details>

```diff
diff --git a/tests/test_lenders.py b/tests/test_lenders.py
index 266a9dc..60956e3 100644
--- a/tests/test_lenders.py
+++ b/tests/test_lenders.py
@@ -1,6 +1,7 @@
 import warnings
 
 from fastapi import status
+from httpx._content import encode_json
 
 from app.models import BorrowerSize, CreditType, UserType
 
@@ -111,26 +112,20 @@ def test_create_credit_product(client):
     response = client.post("/lenders/1/credit-products", json=credit_product, headers=FI_headers)
     assert response.status_code == status.HTTP_401_UNAUTHORIZED
 
-    with warnings.catch_warnings():
-        # "Pydantic serializer warnings" "Expected `decimal` but got `float` - serialized value may not be as expected"
-        warnings.filterwarnings("ignore")
-
-        response = client.post("/lenders/1/credit-products", json=credit_product, headers=OCP_headers)
-        assert response.status_code == status.HTTP_200_OK
+    content_headers, content = encode_json(credit_product)
+    response = client.post("/lenders/1/credit-products", content=content, headers=content_headers | OCP_headers)
+    assert response.status_code == status.HTTP_200_OK
 
     # OCP user tries to create a credit product for a non existent lender
     response = client.post("/lenders/100/credit-products", json=credit_product, headers=OCP_headers)
     assert response.status_code == status.HTTP_404_NOT_FOUND
 
-    with warnings.catch_warnings():
-        # "Pydantic serializer warnings" "Expected `decimal` but got `int` - serialized value may not be as expected"
-        warnings.filterwarnings("ignore")
-
-        response = client.put("/credit-products/1", json=updated_credit_product, headers=OCP_headers)
-        assert response.json()["lower_limit"] == updated_credit_product["lower_limit"]
-        assert response.json()["upper_limit"] == updated_credit_product["upper_limit"]
-        assert response.json()["other_fees_description"] == updated_credit_product["other_fees_description"]
-        assert response.status_code == status.HTTP_200_OK
+    content_headers, content = encode_json(updated_credit_product)
+    response = client.put("/credit-products/1", content=content, headers=content_headers | OCP_headers)
+    assert response.json()["lower_limit"] == updated_credit_product["lower_limit"]
+    assert response.json()["upper_limit"] == updated_credit_product["upper_limit"]
+    assert response.json()["other_fees_description"] == updated_credit_product["other_fees_description"]
+    assert response.status_code == status.HTTP_200_OK
 
     # tries to update a credit product that does not exist
     response = client.put("/credit-products/100", json=updated_credit_product, headers=OCP_headers)
```

</details>

**Comment by yolile** [comment]
Thanks, I will do a final round of testing tomorrow before merging and deploying it

**Comment by yolile** [comment]
> Something weird I had to change is the default value of "source_last_updated_at" in get_award() to None (instead of ""). 

Well, `source_last_updated_at` is a date-time field, not a string, and the default value in the model is None. I guess that is fine and therefore is also fine to keep it with None.

**Comment by jpmckinney** [comment]
Yes, but the JSON data from CCE APIs will be either str or null, and I don’t know what value this field has if it’s empty in the API - so maybe good to double check that

**Comment by yolile** [comment]
> Yes, but the JSON data from CCE APIs will be either str or null

If the value is empty, then it doesn't appear in the JSON response at all

**Comment by yolile** [comment]
Hmm `update_statistics` is failing now with 

`sqlalchemy. exc.ProgrammingError: (psycopg2.errors.UndefinedObject) type "statistictype" does not exist`

It only works if we change the `Statistic` model definition with:

`type: StatisticType = Field(sa_column=Column(Enum(StatisticType, name="statistic_type")))`

But I'm not sure why the other enums are working fine

**Comment by jpmckinney** [comment]
I added a few commits. I think statistictype was causing errors here because the SQLAlchemy calls were a little different, and so SQLAlchemy constructed a query that explicitly used the data type.

**Comment by jpmckinney** [comment]
Note that in 04f931d9ab39b68041e0e9f38817f43aebb5391b I changed created_at to not update. However, we do update statistics, so if the frontend wants to display the `updated_at` (but was using `created_at` in its place), then we should add an `updated_at` field and use it instead.

**Comment by yolile** [comment]
Thanks. Now another problem that we have is that `previous_contracts` is returning a list of empty dicts instead of serializing the database response. I'm trying to figure out why.

**Comment by yolile** [comment]
Note that the `statistictype` issue was already in the main branch and there is also another issue in the main branch as the documents sent by MSMEs are not being saved in the database now (after I deployed the main branch yesterday).

**Comment by yolile** [comment]
> Check whether other endpoints need similar changes as in https://github.com/open-contracting/credere-backend/commit/de5d7423ae39d2b31de42fc3539d9e0ce908c5c6

I've checked, fixed, and tested all the other endpoints (see related commits for details).

> Check if the [recommended approach](https://docs.pydantic.dev/latest/api/standard_library_types/#decimaldecimal) used to fix the above errors ought to be applied to all decimals

I will leave this one for the linked issue.

I will merge this now so that the other issues in production are solved.
